derive_state_machine_future/src/phases.rs

```diff
@@ -304,6 +304,298 @@ impl Pass for ValidPaths {
     }
 }
 
+/// Builds the generics for all states, based on the generics of the state machine.
+#[derive(FromMetaItem, Debug)]
+pub struct BuildStateGenerics;
+
+dummy_default!(BuildStateGenerics);
+
+#[derive(Debug)]
+pub struct BuildStateGenericsExtra {
+    pub generics: Rc<syn::Generics>,
+}
+
+dummy_default!(BuildStateGenericsExtra);
+dummy_from_meta_item!(BuildStateGenericsExtra);
+
+impl Phase for BuildStateGenerics {
+    type StateMachineExtra = <ValidPaths as Phase>::StateMachineExtra;
+    type StateExtra = BuildStateGenericsExtra;
+}
+
+impl Pass for BuildStateGenerics {
+    type FromPhase = ValidPaths;
+
+    fn pass(machine: StateMachine<ValidPaths>) -> StateMachine<BuildStateGenerics> {
+        // returns a Vec that contains (generic_param, bound_ident, idents).
+        // The bound_ident is the ident of the generic_param (can also be a list of idents).
+        // The idents are all idents that will be collected with the CollectIdents trait.
+        fn prepare_generic_params<'a, T, F, R>(
+            param: &'a [T],
+            get_bound: F,
+        ) -> Vec<(&'a T, R, HashSet<syn::Ident>)>
+        where
+            F: Fn(&'a T) -> R,
+            T: CollectIdents,
+        {
+            param
+                .iter()
+                .map(|p| {
+                    let mut idents = HashSet::new();
+                    p.collect_idents(&mut idents);
+
+                    (p, get_bound(p), idents)
+                })
+                .collect::<Vec<_>>()
+        }
+
+        // Checks if a generic_param is part of the state_idents by calling the contains_bound
+        // function. If contains_bound returns true, this function adds all idents that were
+        // collected for this generic_param. For example <T: Clone>, adds ident 'Clone' to
+        // state_idents. Or, <C, T = C> for T, adds ident 'C' to state_idents.
```

Comment:
Adding some <code>`</code>s to demarcate code from prose would be helpful when reading this comment.

Just to check my understanding: the `state_idents` are the idents used by the state either in type/lifetime parameters or bounds of type/lifetime parameters, right? Maybe either spell that out in the comment or rename it to `idents_used` (or even `idents_used_by_state`) or both of those things?